<a id="readme-top"></a>

<br />
<div align="center">
  <h3 align="center">Tumor Microenvironment Research Group</h3>

  <p align="center">
    Projects in Mattias Belting's laboratory at Lund University
    <br />
    <a href="CHANGELOG.md"><strong>Recent changes »</strong></a>
    <br />
  </p>
  <div>
    <div class="row">
        <h4 align="left"> Projects:</h4>
        <div style="display: flex; flex-direction: row; align-items: flex-start;">
            <div style="flex: 33%; padding-right: 10px;">
                <a href="Projects/neo-surfaceome/"><strong>GBM Neosurfaceome</strong></a>
            </div>
            <div style="flex: 67%;">
                <p>The inherent adaptive capacity of GBM to metabolic stress represents a key clinical challenge, and the underlying mechanisms remain ill-defined. Focusing on the emerging role of lipid metabolism in cancer cell stress adaptation, this repository contains the R project contributing to the investigation of mechanistic links between chondrotin sulfate proteoglycan (CSPG) remodelling, extracellular lipid scavenging, and LD formation that together serve to fine tune intracellular lipid metabolism to escape ferroptosis. </p>
            </div>
        </div>
        <br />
        <div style="display: flex; flex-direction: row; align-items: flex-start;">
            <div style="flex: 33%; padding-right: 10px;">
                <a href="Projects/adapted-acidosis/"><strong>Adapted Acidosis</strong></a>
            </div>
            <div style="flex: 67%;">
                <p>Enhanced identification of tumor-specific surface antigens could create new opportunities for immunotherapy, that is yet to be succesfully implemented in GBM therapy. To address the challenge of identifying novel therapeutic vulnerabilities in GBM, our group has developed TS-MAP, a platform for unbiased mapping the tumor surfaceome and endocytome in GBM. This effort classified about 3,300 proteins (SURFME), that shape the tumor's extracellular landscape, identifying several potential therapeutic targets; however, a key limitation emerged – transcriptomic and proteomic data showed limited correlation. Building on these findings, this project seeks to complement TS-MAP by developing an integrative multi-omics pipeline to systematically detect cancer-specific neo-peptides, that arise from either somatic single-nucleotide variants, gene-fusions or alternative splicing, that we collectively termed the neo-surfaceome. </p>
            </div>
        </div>
    </div>
  </div>
</div>